base/iobuffer.jl

```diff
@@ -489,6 +489,86 @@ function take!(io::IOBuffer)
     return data
 end
 
+"Internal method. This method can be faster than takestring!, because it assumes
+the buffer is writable and seekable, and because it does not reset the buffer
+to a usable state.
+Using the buffer after calling unsafe_takestring! may cause illegal behaviour.
+This function is meant to be used when the buffer is only used as a temporary
+string builder, which is discarded after the string is built."
+function unsafe_takestring!(io::IOBuffer)
+    off = io.offset
+    nbytes = io.size - off
+    iszero(nbytes) && return ""
+    # The C function can only copy from the start of the memory.
+    # Fortunately, in most cases, the offset will be zero.
+    return if iszero(off)
+        ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), io.data, nbytes)
+    else
+        mem = StringMemory(nbytes)
+        unsafe_copyto!(mem, 1, io.data, off + 1, nbytes)
+        unsafe_takestring!(mem)
+    end
+end
+
+"""
+    takestring!(io::IOBuffer) -> String
+
+Return the content of `io` as a `String`, resetting the buffer to its initial
+state.
+This is preferred over calling `String(take!(io))` to create a string from
+an `IOBuffer`.
+
+# Examples
+```jldoctest
+julia> io = IOBuffer();
+
+julia> write(io, [0x61, 0x62, 0x63]);
+
+julia> s = takestring!(io)
+"abc"
+
+julia> isempty(take!(io)) # io is now empty
+true
+```
+
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
+"""
+function takestring!(io::IOBuffer)
+    nbytes = filesize(io)
+
+    # If the memory is invalidated (and hence unused), or no bytes, return empty string
+    return if (iszero(nbytes) || io.reinit)
+        ""
+    else
+        start = io.seekable ? io.offset + 1 : io.ptr
```

Comment:
Why is this offset computed differently in `unsafe_takestring!`?

Reply:
See above.